<font size = 5>Collect the Data<Br><font color = 'blue'><i>Obtener los Datos</i></font></font><br>https://github.com/nmolivo/valencia-data-projects

<font size = 3><b>Use 'Requests' library to save a JSON url.</b></font><font size = 2 color = 'blue'><br><i>Usar Python librería 'Requests' a guardar los archivos JSON de una URL.</i></font><br>

A scraper used to automate downloading Spain Valenbisi data from the following website:<font size = 2 color ='blue'><i><br> Un rascador del web por la búsqueda de los datos del España Valenbisi:</i></font><br><br>
http://www.ayto-valencia.es/ayuntamiento/datosabiertos.nsf/resultadoCapas/95D6756F6861F9FDC1257C70003E4FBD?OpenDocument&lang=1&nivel=2&seccion=1&bdorigen=&idapoyo=22ADF97C1FD223B5C1257C55003BD01F<br><br>

<img src="http://www.ayto-valencia.es/ayuntamiento/DatosAbiertos.nsf/0/8A266EF774003DC0C1257C55003BDD87/$FILE/VDatosabiertos.jpg">

<b>1.</b>
<br>Import the libraries 
<font color = 'blue'>Importa los paquetes</font>

If you encounter an error, go to command line:<br><font color = 'blue'>Si encuentra un error ve al símbolo del sistema:</font><Br>
`>> pip install requests`

In [37]:
import pandas as pd

import requests
import json

<b>2.</b><br>Request data<br>
Documentation: http://docs.python-requests.org/en/master/<br><br>
<font color = 'blue'>Solicita los datos<Br>
Documentation:</font> https://pythonbc.com/blog/la-libreria-requests-de-python/</font>

In [90]:
r = requests.get("http://mapas.valencia.es/lanzadera/opendata/Valenbisi/JSON")

<b>3.</b><br>Confirm the connection<br>
HTTP response status codes: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status
<font color = 'blue'>Constata la conexión<br>
Códigos de estado de respuesta HTTP:</font> https://developer.mozilla.org/es/docs/Web/HTTP/Status

In [91]:
r.status_code

200

<b>4.</b><br>
Write `'w'` and Read `'r'` the file we downloaded from the URL<br>We use `with`-statements so we do not have to manually close the file.<br>Otherwise, we would need to do: <br>
~~~
f.open('filename')
json.dump(res, f)
f.close()
~~~
<font color = 'blue'>Escribir `'w'` y Leer `'r'` el archivo nos descargamos del URL<br>
Usamos `with`-declaraciónes a fin de que no necesitamos cerrar manualmente el archivo.<br>De otra manera, necesitamos que hacer:</font>
~~~
f.open('filename')
json.dump(res, f)
f.close()
~~~

In [92]:
res = r.json()

In [93]:
with open('valenbici.json', 'w') as f:
    json.dump(res, f, ensure_ascii=False)

In [94]:
with open('valenbici.json', 'r') as f:
    data=json.load(f)

<b>5.</b><br>
Explore the format of our output.<br><font color = 'blue'>Exploremos el formato de los datos.

In [113]:
data
#We have nested dictionaries.
#Tenemos diccionarios encajados.

{'crs': {'properties': {'name': 'urn:ogc:def:crs:EPSG::25830'},
  'type': 'name'},
 'features': [{'geometry': {'coordinates': [726674.382, 4370336.97],
    'type': 'Point'},
   'properties': {'address': 'Bombero Ramón Duart - Hermanos Maristas',
    'available': '6',
    'free': '8',
    'name': '184_CALLE_BOMBER_RAMON_DUART',
    'number': '184',
    'open': 'T',
    'ticket': 'F',
    'total': '15',
    'updated_at': '20/02/2018 01:03:55'},
   'type': 'Feature'},
  {'geometry': {'coordinates': [724673.385, 4369442.063], 'type': 'Point'},
   'properties': {'address': 'Plaza Salvador Soria, 8',
    'available': '1',
    'free': '19',
    'name': '257_PLAZA_SALVADOR_SORIA_ESQ_PIO_X',
    'number': '257',
    'open': 'T',
    'ticket': 'F',
    'total': '20',
    'updated_at': '20/02/2018 01:03:55'},
   'type': 'Feature'},
  {'geometry': {'coordinates': [726975.018, 4373093.228], 'type': 'Point'},
   'properties': {'address': 'Blasco Ibañez, 28 (F. Geografía e Historia)',
    'available'

In [114]:
data.keys()
#A way to see the keys. Dictionaries also have the property .values()
#Un modo para ver los keys. Diccionarios tiene la característica .values() tambien.

dict_keys(['features', 'crs', 'type'])

In [117]:
data['type']
#What information is in 'type'? 'crs'? Not much. The important information is in 'features'
#Qué información es en 'type'? 'crs'? Nada. Los información importante es en 'features'

'FeatureCollection'

In [97]:
len(data['features'])
#How many items are in our data? The same number as there are Valenbisi stations
#Quantos ítems en nuestros datos? Lo mismo que la cantidad de los estaciones de Valenbisi.

273

In [99]:
data['features'][0]
#features keys: 'properties', 'type', 'geometry'

{'geometry': {'coordinates': [726674.382, 4370336.97], 'type': 'Point'},
 'properties': {'address': 'Bombero Ramón Duart - Hermanos Maristas',
  'available': '6',
  'free': '8',
  'name': '184_CALLE_BOMBER_RAMON_DUART',
  'number': '184',
  'open': 'T',
  'ticket': 'F',
  'total': '15',
  'updated_at': '20/02/2018 01:03:55'},
 'type': 'Feature'}

<b>6.</b><br>Define empty lists for each column in our DataFrame. We will populate these lists using our data.<br><font color='blue'>Definir las listas vacías por cada columna en la DataFrame. Los llenaremos las listas con los datos.</font>

In [101]:
X_values = []
Y_values = []
name_values = []
number_values = []
open_values = []
ticket_values = []
available_values = []
free_values = []
total_values = []
update_values = []

<b>7.</b><br>Iterate through each item in our data (273) and save the information in our empty lists.<br><font color = 'blue'>Iterar a cada ítem en nuestros datos (273) y guardar la informacion en las listas vacías.</font>

In [102]:
for record in range(0, len(data['features'])):
    xx = data['features'][record]['geometry']['coordinates'][0]
    yy = data['features'][record]['geometry']['coordinates'][1]
    nombre = data['features'][record]['properties']['name']
    numero = data['features'][record]['properties']['number']
    abierto = data['features'][record]['properties']['open']
    boleta = data['features'][record]['properties']['ticket']
    disponible = data['features'][record]['properties']['available']
    feliz = data['features'][record]['properties']['free']
    cantidad = data['features'][record]['properties']['total']
    fecha = data['features'][record]['properties']['updated_at']
    
    X_values.append(xx)
    Y_values.append(yy)
    name_values.append(nombre)
    number_values.append(numero)
    open_values.append(abierto)
    ticket_values.append(boleta)
    available_values.append(disponible)
    free_values.append(feliz)
    total_values.append(cantidad)
    update_values.append(fecha)

<b>8.</b><br>Define a dictionary with column name as `keys` and our lists of data as `values`.<br><font color = 'blue'>Definir una diccionario con los nombres de las columnas como `keys` y las listas de datos como `values`.
~~~
diccionario{key_1 : value_1,
            key_2 : value_2,
            .
            .
            .
            key_n : value_n
            }
~~~

In [103]:
bici_dict = {"X":X_values,
            "Y":Y_values,
            "name":name_values,
            "number":number_values,
            "open": open_values,
            "ticket":ticket_values,
            "available":available_values,
            "free":free_values,
            "total":total_values,
            "update":update_values}

<b>9.</b><br>
Convert the dictionary to a DataFrame.<br><font color = 'blue'>Convirtir la diccionario a una DataFrame.</font>

In [87]:
df = pd.DataFrame(bici_dict)

<b>10.</b><br>Save the DataFrame to our computers.<br><font color = 'blue'>Guardar la DataFrame a nuestras computadoras.</font>

In [89]:
df.to_json('valencia_bici.json', force_ascii=False)

In [104]:
df

X            Y available free  \
0    727325.931  4372577.687        17    3   
1    726887.830  4372671.832         5   20   
2    726996.881  4372829.010         2   18   
3    726473.430  4373086.657         0   25   
4    729243.245  4372024.795        15    0   
5    726661.312  4373300.906         1   22   
6    724762.450  4371533.357         8   12   
7    726121.820  4374701.859        15    5   
8    726591.991  4370633.414        17    6   
9    726223.891  4370683.656        10   10   
10   726431.916  4370909.480         7    8   
11   726097.133  4370302.116        13    5   
12   726674.382  4370336.970         6    8   
13   726258.742  4370020.644         8    7   
14   726410.505  4369793.547         1   14   
15   725993.732  4369396.438         0   40   
16   725811.479  4369410.682         0   30   
17   725428.757  4369556.305         7   11   
18   725543.743  4370287.886         7   13   
19   724846.881  4370546.384        16    4   
20   725322.215  4371212.544         1   39   
21   725042.932  4371022.899         7    9   
22   724353.939  4370869.611        11    4   
23   724088.606  4371161.449         2   18   
24   724409.686  4371249.394        15    0   
25   723879.043  4372250.791         7    8   
26   723510.284  4372471.323         6   12   
27   723788.443  4372594.236         4   12   
28   723394.032  4372740.052         3   17   
29   723526.821  4373109.163        16    2   
..          ...          ...       ...  ...   
243  729293.988  4371599.114        17    0   
244  730145.464  4372181.084        17    0   
245  729837.887  4373106.430        14    0   
246  724192.082  4373451.012         2   18   
247  725404.568  4374658.835        15    5   
248  725861.045  4374841.896        14    1   
249  725857.325  4370607.531        10   11   
250  725955.382  4370050.348         3   11   
251  725538.115  4370031.311         6    9   
252  724716.630  4371034.651        12    3   
253  724247.446  4371927.825         1   14   
254  723899.147  4373929.248         4   16   
255  724314.852  4370190.576         5   13   
256  723623.032  4371598.159         3   12   
257  723259.895  4372187.155         2   22   
258  723868.082  4372934.044        19    1   
259  723339.372  4374402.457         2   18   
260  726522.722  4374983.326        10    5   
261  724051.437  4375223.032         6    9   
262  725009.401  4371403.652        18    0   
263  725016.729  4372512.385         4   11   
264  727032.722  4371091.257        10   10   
265  727210.757  4372259.455         6   14   
266  725020.763  4372874.455         7    8   
267  723101.260  4373091.117         8   12   
268  723251.749  4371355.044         2   18   
269  723616.945  4370241.960        13    2   
270  724336.117  4369455.518        17    2   
271  722836.087  4371810.010        13    5   
272  724431.858  4375437.359         4   11   

                                                  name number open ticket  \
0                                  079_AVDA. ARAGON II     79    T      F   
1                                    080_AMADEO SAVOIA     80    T      F   
2                                    081_DOCTOR FORNOS     81    T      F   
3                        83_PZA. DE LA LEGIÓN ESPAÑOLA     83    T      T   
4              84_SERRERIA_ESQ_PINTOR FERRER CALATAYUD     84    T      F   
5                             085_AVDA_BLASCO IBAÑEZ 1     85    T      F   
6                             157_AVENIDA_PEREZ_GALDOS    157    T      T   
7                               178_CALLE_REIG_GENOVES    178    T      F   
8                           179_AVENIDA_DE_LA_PLATA_45    179    T      F   
9                           180_AVENIDA_DOCTOR_WAKSMAN    180    T      T   
10                               181_CALLE_DEL_SALINAR    181    T      T   
11                               183_AVDA_AUSIAS_MARCH    183    T      T   
12                        184_CALLE_BOMBER_RAMON_DUART    184    T      F   
13                   

In [4]:
"%05d" % (10,)

'00010'